### PR TITLE
PHP-1238: Support authMechanismProperties MongoClient option

### PR DESCRIPTION
**Pull request by jmikola**
https://jira.mongodb.org/browse/PHP-1238

### CODE REVIEW

File: mcon/parse.c [hunk 1]
```diff
@@ -285,6 +285,54 @@ static int mongo_process_option(mongo_con_manager *manager, mongo_servers *serve
...
+	do {
+		end = strchr(start, ',');
+		colon = strchr(start, ':');
+
+		if (!colon) {
+			int len = strlen(start) + sizeof("Error while trying to parse auth mechanism properties: No separator for ''");
+
+			*error_message = malloc(len + 1);
+			snprintf(*error_message, len, "Error while trying to parse auth mechanism properties: No separator for '%s'", start);
+			return 3;
```
Comment (derickr) [inline mcon/parse.c]: You need to document the return values in the pre-function comment.

File: mcon/parse.c [hunk 2]
```diff
@@ -285,6 +285,54 @@ static int mongo_process_option(mongo_con_manager *manager, mongo_servers *serve
...
+
+	do {
+		end = strchr(start, ',');
+		colon = strchr(start, ':');
+
+		if (!colon) {
+			int len = strlen(start) + sizeof("Error while trying to parse auth mechanism properties: No separator for ''");
+
+			*error_message = malloc(len + 1);
+			snprintf(*error_message, len, "Error while trying to parse auth mechanism properties: No separator for '%s'", start);
```
Comment (derickr) [inline mcon/parse.c]: Kinda annoying to have the string in here twice - but I guess in other places we didn't make a constant either?

File: mcon/parse.c [hunk 3]
```diff
@@ -285,6 +285,54 @@ static int mongo_process_option(mongo_con_manager *manager, mongo_servers *serve
...
+	start = value;
+
+	do {
+		end = strchr(start, ',');
+		colon = strchr(start, ':');
+
+		if (!colon) {
+			int len = strlen(start) + sizeof("Error while trying to parse auth mechanism properties: No separator for ''");
+
+			*error_message = malloc(len + 1);
```
Comment (derickr) [inline mcon/parse.c]: I probably would have prefixed the "strlen(start)" two lines above with the "1 + " --- i think that's more common.

File: mcon/parse.c [hunk 4]
```diff
@@ -285,6 +285,54 @@ static int mongo_process_option(mongo_con_manager *manager, mongo_servers *serve
...
+			snprintf(*error_message, len, "Error while trying to parse auth mechanism properties: No separator for '%s'", start);
+			return 3;
+		}
+
+		prop_name = mcon_strndup(start, colon - start);
+		prop_value = mcon_strndup(colon + 1, end - colon - 1);
+
+		if (strcasecmp(prop_name, "SERVICE_NAME") == 0) {
+			mongo_manager_log(manager, MLOG_PARSE, MLOG_INFO, "- Found auth mechanism property '%s': '%s'", prop_name, prop_value);
+			free(servers->options.gssapiServiceName);
```
Comment (derickr) [inline mcon/parse.c]: Shouldn't you check whether it is set before you free it? free(null) doesn't work well on every platform.

File: mcon/parse.c (hunk 2 above)
Comment (jmikola) [inline mcon/parse.c]: I also found it odd that we `malloc()` to the exact length here, but elsewhere (e.g. for unsupported options) use a constant buffer of 256 and allow the error message to be truncated.

In any event, I'm not sure a constant for the error message would help unless we want to define it with `"%s"` included and subtract 2 when calculating the length.

File: mcon/parse.c (hunk 4 above)
Comment (jmikola) [inline mcon/parse.c]: The option value is initialized to `strdup("mongodb")` and is only ever assigned to a newly allocated string thereafter. This was copied from the existing option parsing of `gssapiServiceName`.

File: mcon/parse.c (hunk 3 above)
Comment (jmikola) [inline mcon/parse.c]: Just noticed that every other `malloc()` in `parse.c` does the above and adds `+ 1`. Going to leave this as-is for consistency.

File: mcon/parse.c (hunk 1 above)
Comment (jmikola) [inline mcon/parse.c]: Done.

Comment (derickr) [inline mcon/parse.c]: You do a "Return 2" too (line 339).

**Comment by derickr** [comment]
LGTM, once you:

 - [ ] Document the ```return 2;``` too.

File: mcon/parse.c (hunk 1 above)
Comment (jmikola) [inline mcon/parse.c]: Ah, good catch. Fixed.

**Pull request merged**